### ISSUE TITLE
Type changing traversals?

### ISSUE THREAD

**Issue by ocharles** [sjoerdvisscher/one-liner#1]
Excuse the terrible title, I'm not sure how to exactly refer to this operation. Currently in my `rel8` project, I've had to manually define a fairly boring generic traversal:

```haskell
-- | Map a schema definition into a set of expressions that would select those
-- columns.
class InferBaseTableAttrExpr schema expr where
  baseTableAttrExpr :: schema a -> expr a

instance (InferBaseTableAttrExpr schema expr) =>
         InferBaseTableAttrExpr (M1 i c schema) (M1 i c expr) where
  baseTableAttrExpr (M1 s) = M1 (baseTableAttrExpr s)

instance ( InferBaseTableAttrExpr fSchema fExpr
         , InferBaseTableAttrExpr gSchema gExpr
         ) =>
         InferBaseTableAttrExpr (fSchema :*: gSchema) (fExpr :*: gExpr) where
  baseTableAttrExpr (l :*: r) = baseTableAttrExpr l :*: baseTableAttrExpr r

instance InferBaseTableAttrExpr (K1 i (Tagged name String)) (K1 i (Expr a)) where
  baseTableAttrExpr (K1 (Tagged name)) = K1 (Expr (O.BaseTableAttrExpr name))
```

It is used as such:

```haskell
  tableDefinition :: ( ADTRecord (table Expr)
                       , ADTRecord (table Schema)
                       , Constraints (table Schema) WitnessSchema
                       , InferBaseTableAttrExpr (Rep (table Schema)) (Rep (table Expr))
                       , Writer (Rep (table Schema)) (Rep (table Insert))
                       , Generic (table Insert)
                       )
                    => O.Table (table Insert) (table Expr)
  tableDefinition =
    O.Table
         (symbolVal (Proxy :: Proxy name))
         (O.TableProperties
            (case lmap from (columnWriter (from tableSchema)) of
               O.Writer f -> O.Writer f)
            (O.View (to (baseTableAttrExpr (from tableSchema)))))
    where
      tableSchema :: table Schema
      tableSchema = nullaryOp (For :: For WitnessSchema) schema
```

Notice the use of `nullaryOp` to construct the schema, which is then transformed back into its representation, passed to `baseTableAttrExpr`, and then the whole thing is moved back into concrete types.

**Comment by sjoerdvisscher** [comment]
Ah, I was thinking, if I want to take this idea to its full generality, I should use type classes with 2 parameters to be able to change types at the K1 leafs. But I couldn't think of a good usecase for it (the ones I could think of were fixed with the generics1 versions) so I abandonned the idea. But here you are!

I don't think I will add this soon, everything became quite messy with two parameters everywhere. The only way around this for now is that somehow you turn this into a Generic1 case, with the parameter being the only place where actually can change types in the upcoming version.

**Comment by sjoerdvisscher** [comment]
Here's another example that could use this: https://github.com/tomjaguarpaw/product-profunctors/pull/19/files

**Comment by sjoerdvisscher** [comment]
Implemented in version 1.0!

**Comment by ocharles** [comment]
Oh wow, terrific! I'll have a play :smile:
